### PR TITLE
Update rubygems

### LINKED ISSUE #36: Update rubygems as part of bootstrap
Rubygems should be updated for homebrew Ruby.

>  gem update --system --no-document

### PR DESCRIPTION
**Pull request by bootstraponline**
### New Pull Request Checklist

*Please mark the points which you did / accept.*

- [x] The tool I added is stable, and __does not require frequent updates__. _Preinstalled tools on the LTS stacks
  are not updated, so if the tool requires frequent updates you should handle the installation on-demand (see: [Install Any Additional Tool - DevCenter](https://bitrise-io.github.io/devcenter/tips-and-tricks/install-additional-tools/) for more information)_
- [x] I added a version report line to [`system_report.sh`](/system_report.sh) for the new tool(s) I added

Fix #36

### CODE REVIEW

**Comment by viktorbenei** [comment]
Thanks again @bootstraponline !

Unfortunately as this week's updates are now finalized this update can only be part of next week's updates - but it will be! ;)

**Comment by bootstraponline** [comment]
I already have this in my fastlane so no worries. 😄

**Comment by viktorbenei** [comment]
😉

**Comment by viktorbenei** [comment]
Will be included in this weekend's VM update (Feb 25th)

**Comment by viktorbenei** [comment]
Unfortunately we had to remove the rubygems update.

The issue is: if you don't do this, and just rely on the MacOS preinstalled rubygems, you can simply `sudo gem install ...` gems which install bins/executables. Simple example: `sudo gem install bundler`.

Once you do `sudo gem update --system --no-document` the path to gem bins changes to `/usr/bin/`, which is off limit since El Capitan / SIP and you can't write to that even with `sudo`.

Of course the Ruby recommendation is to not use `sudo` to install gems, but we can't prevent others to do that, especially because lots of devs use the MacOS system default Ruby version and write scripts which rely on that.

You can btw reproduce this on a brand new MacOS install, and also on any other MacOS Sierra if you uninstall the non system rubys first, then you do:

```
# uninstall the non system ruby, if any
brew uninstall ruby
# rubygems update
sudo gem update --system --no-document
# try to install any gem which installs a bin/executable
sudo gem install bundler
```

You'll get this error:

```
+ sudo gem install bundler
ERROR:  While executing gem ... (Errno::EPERM)
    Operation not permitted - /usr/bin/bundle
```

We'll try to find a better solution, maybe by specifying an alternative "EXECUTABLE DIRECTORY" for Ruby with an environment variable - although as a general guideline we try to keep the Bitrise environment as close to a standard developer Mac environment as possible, changing as few things as possible, to avoid hard to debug issues caused by things like your scripts run fine on your Mac, expecting `sudo gem install X` to succeed and install the executable to dir Y, while on Bitrise it'll install it to somewhere else or won't install it at all with `sudo`.

It's quite a tricky issue, we're not sure why `sudo gem update --system --no-document` changes the executable directory of ruby at all.. But we tested this on a brand new, freshly installed Sierra, and it can also be tested with a simple Bitrise.io build after `brew uninstall ruby` to only have the system installed Ruby available:

```
+ sudo gem env
RubyGems Environment:
  - RUBYGEMS VERSION: 2.0.14.1
  - RUBY VERSION: 2.0.0 (2015-12-16 patchlevel 648) [universal.x86_64-darwin16]
  - INSTALLATION DIRECTORY: /Library/Ruby/Gems/2.0.0
  - RUBY EXECUTABLE: /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby
  - EXECUTABLE DIRECTORY: /usr/local/bin
  - RUBYGEMS PLATFORMS:
    - ruby
    - universal-darwin-16
  - GEM PATHS:
     - /Library/Ruby/Gems/2.0.0
     - /Users/vagrant/.gem/ruby/2.0.0
     - /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/gems/2.0.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/

+ sudo gem update --system --no-document
Updating rubygems-update
Successfully installed rubygems-update-2.6.11
Installing RubyGems 2.6.11
RubyGems 2.6.11 installed
....
RubyGems installed the following executables:
	/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/gem

RubyGems system software updated

+ sudo gem env
RubyGems Environment:
  - RUBYGEMS VERSION: 2.6.11
  - RUBY VERSION: 2.0.0 (2015-12-16 patchlevel 648) [universal.x86_64-darwin16]
  - INSTALLATION DIRECTORY: /Library/Ruby/Gems/2.0.0
  - USER INSTALLATION DIRECTORY: /Users/vagrant/.gem/ruby/2.0.0
  - RUBY EXECUTABLE: /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby
  - EXECUTABLE DIRECTORY: /usr/bin
  - SPEC CACHE DIRECTORY: /Users/vagrant/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /Library/Ruby/Site
  - RUBYGEMS PLATFORMS:
    - ruby
    - universal-darwin-16
  - GEM PATHS:
     - /Library/Ruby/Gems/2.0.0
     - /Users/vagrant/.gem/ruby/2.0.0
     - /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/gems/2.0.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /Users/vagrant/.bitrise/tools
     - /usr/local/bin
     - /usr/local/sbin
     - ~/bin
     - /usr/local/bin
     - /usr/bin
     - /bin
     - /usr/sbin
     - /sbin
     - /usr/local/opt/go/libexec/bin
     - /Users/vagrant/go/bin
     - /Users/vagrant/bitrise/tools/cmd-bridge/bin/osx
```

Notice the change of "EXECUTABLE DIRECTORY" from the first/default `/usr/local/bin`, where `sudo gem install` can install, to the `/usr/bin` path where it can't since El Capitan/SIP.

We're still trying to figure out the "right" way, which would work for everyone. If you have any idea @bootstraponline we're always happy to discuss!

Of course you can still run `sudo gem update --system --no-document` with e.g. a __Script__ step, if you don't want to use the system ruby / `sudo gem install` after that (which is a good thing, to not to `sudo gem install`, but lots of devs do `sudo gem install`) and you need the updated rubygems.

**Comment by bootstraponline** [comment]
I think the right way would be to leave system ruby untouched. Then install the latest Ruby via rbenv or similar with the correctly updated rubygems. What do you think?

**Comment by bootstraponline** [comment]
also have you tried `gem: --bindir /usr/local/bin` in `~/.gemrc`?

**Comment by viktorbenei** [comment]
> I think the right way would be to leave system ruby untouched

I definitely agree.

> Then install the latest Ruby via rbenv or similar with the correctly updated rubygems.

Also agree, we just opt to use the brew install ruby version instead of rbenv or rvm, because Xcode's command line tools did have issues with RVM/rbenv installed rubies in the past.

> What do you think?

Totally agree, we do this right now, but the issue is that `sudo gem update --system --no-document` does change the system ruby, regardless of what ruby you have installed and from where. It changes the MacOS default ruby's `EXECUTABLE DIRECTORY` for some unexplained reason..

There might be a way to prevent it to change the `EXECUTABLE DIRECTORY`, or we could try to use something like what you mentioned (`~/.gemrc`), but we'll still have to see if the `EXECUTABLE DIRECTORY` is the only one which can cause issues, or `sudo gem update --system --no-document` changes something else too which can cause problems when you `sudo gem install ..`

**Comment by bootstraponline** [comment]
I wonder what the other CI providers (Travis CI) do to solve this issue. If we're only running the update command on Homebrew Ruby then we shouldn't need sudo.

**Comment by viktorbenei** [comment]
> I wonder what the other CI providers (Travis CI)

Depends. Most of the traditional CIs started from a web background, so e.g. Travis do use RVM. That said, Travis does not do `sudo gem update --system`, at least that's not in their bootstrap script: https://github.com/travis-infrastructure/osx-image-bootstrap/blob/master/bootstrap.sh

If you don't do `sudo gem update --system` there's no issue, regardless of what Ruby version manager (if any) you use.

**Comment by bootstraponline** [comment]
That's interesting. I guess the `~/.gemrc` solution would be the next thing to try then. It seems strange that we can have isolated Ruby versions but rubygems is messing things up globally.

**Comment by viktorbenei** [comment]
>  It seems strange that we can have isolated Ruby versions but rubygems is messing things up globally.

Well, the thing is, if you do `sudo` (!) on MacOS you won't run the command for the isolated/managed Ruby, that will always be for the system preinstalled Ruby. E.g. `~/.gemrc` will only affect the user who's home dir you save it into. When you run `sudo ...` you effectively switch to the __root__ user for the duration of the command, and the root user has its own home dir, envs etc.

**Comment by bootstraponline** [comment]
Why can't we just use `gem update --system --no-document` without sudo?

**Comment by viktorbenei** [comment]
We definitely can, if it's enough to update it for the default (brew installed) ruby version. At least it should be - we only tested `sudo gem update ..` as we wanted to update it for the system ruby too.

**Comment by bootstraponline** [comment]
I think we should leave system ruby untouched & not upgraded.

**Comment by viktorbenei** [comment]
Yeah, it seemed a good idea initially, until we hit this strange executable dir path change issue with `sudo gem update --system`

**Comment by viktorbenei** [comment]
One more question though, whether we should do `gem update --system` every week on every stack, or should only be performed as part of the initial provision of a stack and not updated after that.

I'd vote for only in initial provisioning, as the purpose of Long Term Support / Stable stacks is to have the least amount of change once the stack is prepared, to provide maximum stability.

For reference this is what we run on every stack every week: https://github.com/bitrise-io/osx-box-bootstrap/blob/master/weekly-cache-update-playbook.yml

TL;DR;

- `brew update` (but no upgrade - this just refreshes brew's local db/spec)
- `pod setup` (again, this will only refresh cocoapods' local db/spec)
- Bitrise CLI upgrade to the latest version
- and a couple of other infrastructure related Bitrise tools upgrades to the latest versions
- DerivedData cleanup (for historical reasons, as we used to run test builds directly as part of the provisioning)

**Comment by bootstraponline** [comment]
> should only be performed as part of the initial provision of a stack and not updated after that.

how often does initial provision happen?

If I'm on an edge stack then I'm not seeing a benefit to running obsolete software. For LTS / Stable, sure don't update it weekly. It's not clear if Linux has an "edge" stack though.

**Comment by viktorbenei** [comment]
> how often does initial provision happen?

Only once in case of LTS stacks, weekly for "rolling release / edge" stacks. E.g. the Xcode 7.3 stack was provisioned only once, and since that we only ran this weekly cache update on it, nothing else changed (we do these cache updates with cloned images, so that we can roll it back if we have to).

> If I'm on an edge stack then I'm not seeing a benefit to running obsolete software.

Sure, Edge is re-built from the ground up every week, including initial provisioning.

> For LTS / Stable, sure don't update it weekly.

I agree, probably best to do this only as part of the initial provisioning.

> It's not clear if Linux has an "edge" stack though.

The default Android/Linux is the same as "edge", rebuilt every week. Although in case of Linux it's a bit easier, as there's no system default ruby.

More info about stack & update types: https://discuss.bitrise.io/t/how-are-the-bitrise-io-stacks-updated-and-with-what-frequency/281

**Comment by bootstraponline** [comment]
ok, all that sounds awesome. +1 from me.

**Comment by viktorbenei** [comment]
Created a new PR - https://github.com/bitrise-io/osx-box-bootstrap/pull/39

BTW part of the issue was `sudo gem update ..`, the other reason why we removed this was related to brew, where they keep changing how they handle versioning of tools/formulas, and they broke how we previously locked the default Ruby version to 2.3 (we did lock it to 2.3 because of the breaking issues ruby based tools had initially when Ruby 2.4 came out last year).

Now that we could switch back to the "default" brew ruby (2.4 right now) this should work again.

**Comment by viktorbenei** [comment]
Man I love Ruby & ruby based tools.. Life would be so boring without it :D
